gax-httpjson/src/main/java/com/google/api/gax/httpjson/ResourceNameStruct.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2018, Google LLC All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.resourcenames.ResourceName;
+import java.util.Map;
+
+/* A ResourceName that is essentially a dictionary mapping of fieldNames to values. */
+public interface ResourceNameStruct extends ResourceName {
+  /* Fetch the comprehensive mapping of fieldNames to values. The set of keys of the resulting Map
+  should be constant, for a given instance of this interface. */
+  Map<String, String> getFieldValuesMap();
+
+  /* Return a new instance of this interface by parsing a formatted String. This should be treated as a static method. */
+  ResourceNameStruct parseFrom(String formattedString);
+}
```

Comment:
`getFieldValuesMap()` will be unnecessary once https://github.com/googleapis/api-common-java/pull/46 is rolled out.

After that, the only thing then left is `parseFrom`. It appears that this is actually an attempt to do an abstract factory, but using a dummy instance of the type being created. I think it would be more natural to create an actual abstract factory. This means:
  - Create a `ResourceNameFactory` interface in api-common-java, parameterized with the resource name type (`ResourceNameT`), add `parseFrom` to it
  - Update the generated resource name classes to generate a method like so:

```
public class AddressName implements ResourceName {
  public static ResourceNameFactory<AddressName> newFactory() {
    return new ResourceNameFactory<AddressName>() {
      public AddressName parseFrom(String formattedString) {return AddressName.parseFrom(formattedString);}
    };
  }
}
```

  - Then, pass a `ResourceNameFactory` to the constructor of `ApiMessageHttpRequestFormatter`. 
  - Finally, delete `ResourceNameStruct`.

Reply:
Sounds awesome! Done.